FlashLFQ/FlashLfqEngine.cs

```diff
@@ -519,55 +526,241 @@ private void QuantifyMatchBetweenRunsPeaks(SpectraFileInfo idAcceptorFile)
+                                }
+                            }
+
+                            // used for calculating FDR
+                            bool isKnownPeak = sequenceToRealPeakInAcceptorFile.ContainsKey(identification.ModifiedSequence);
+                            if (isKnownPeak)
+                            {
+                                // get the "real" peak in this file
+                                var realPeak = sequenceToRealPeakInAcceptorFile[identification.ModifiedSequence];
+
+                                if (realPeak == null || realPeak.Intensity == 0)
+                                {
+                                    matchedPeaks[i] = null;
+                                    continue;
+                                }
+
+                                bool isCorrectMatch = realPeak.Apex.IndexedPeak.Mz == bestAcceptorPeak.Apex.IndexedPeak.Mz
+                                    && realPeak.Apex.IndexedPeak.RetentionTime == bestAcceptorPeak.Apex.IndexedPeak.RetentionTime;
+
+                                scoreWithKnownResults[i] = (bestAcceptorPeak, isCorrectMatch);
+                                matchedPeaks[i] = null;
+                            }
                         }
                     });
 
-                // save MBR results
-                foreach (var peak in matchedPeaks.Where(p => p != null && p.IsotopicEnvelopes.Any()))
+                var res = scoreWithKnownResults.Where(p => p.bestAcceptorPeak != null).OrderBy(p => p.bestAcceptorPeak.MbrScore).ToList();
+
+                int cumulativeTruePositives = res.Count(p => p.isTruePositive);
+                int cumulativeFalsePositives = res.Count(p => !p.isTruePositive);
+                double scoreCutoff = 1;
+
+                foreach (var (bestAcceptorPeak, isTruePositive) in res)
+                {
+                    double qValue = (double)cumulativeFalsePositives / (cumulativeFalsePositives + cumulativeTruePositives);
+
+                    if (qValue <= MbrQValueCutoff)
+                    {
+                        scoreCutoff = bestAcceptorPeak.MbrScore;
+                        break;
+                    }
+
+                    if (isTruePositive)
+                    {
+                        cumulativeTruePositives--;
+                    }
+                    else
+                    {
+                        cumulativeFalsePositives--;
+                    }
+                }
+
+                // calculate FDR
+                foreach (var peak in matchedPeaks.Where(p => p != null))
+                {
+                    int cumulativeTrue = res.Count(p => p.isTruePositive && p.bestAcceptorPeak.MbrScore >= peak.MbrScore);
+                    int cumulativeFalse = res.Count(p => !p.isTruePositive && p.bestAcceptorPeak.MbrScore >= peak.MbrScore);
+                    double qValue = ((double)cumulativeFalse) / (cumulativeTrue + cumulativeFalse);
+
+                    if (cumulativeFalse == 0)
```

Comment:
Is this needed? If cumulativeFalse is zero, then q-value should also be zero (unless cumulativeTrue is also zero, but then we don't have anything and we wouldn't be in this loop...)